### PR TITLE
Implementation for FTP, FTPS and SFTP components in SIP Test Kit

### PR DESCRIPTION
**Pull request by nikolag-ikor**
# Description

Implementation for supporting FTP, FTPS and SFTP in SIP Test Kit.

## Type of change

Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing (regression) unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: sip-test-kit/pom.xml [hunk 1]
```diff
@@ -73,5 +73,13 @@
                 </exclusion>
             </exclusions>
         </dependency>
+        <dependency>
+            <groupId>org.apache.camel.springboot</groupId>
+            <artifactId>camel-ftp-starter</artifactId>
```
Comment (vladiIkor) [inline sip-test-kit/pom.xml:78]: I would have this dependency as optional

File: sip-test-kit/pom.xml [hunk 2]
```diff
@@ -73,5 +73,13 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.camel</groupId>
```
Comment (vladiIkor) [inline sip-test-kit/pom.xml]: Why do we need this one?

File: sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    }
+  }
+
+  private void prepareHeadersWithFilename(
+      Map<String, Object> headers, String endpointAbsolutePath) {
+    String filename =
+        normalizePathToProtocol(
+            FileUtil.stripLeadingSeparator((String) headers.get(CAMEL_FILE_NAME.getValue())));
+    headers.put(CAMEL_FILE_NAME.getValue(), filename);
+    if (!headers.containsKey(CAMEL_FILE_NAME_CONSUMED.getValue())) {
```
Comment (vladiIkor) [inline sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java]: Looks like you need headers.putIfAbsent(key, value)

Comment (nikolag-ikor) [reply sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java]: Resolved

File: sip-test-kit/pom.xml (hunk 2 above)
Comment (nikolag-ikor) [reply sip-test-kit/pom.xml]: There is a camel FileUtil class coming from that dependency. Methods from FileUtil class are used for setting some of FTP exchange headers in our FtpRouteInvoker. Same behaviour for setting headers is in FtpComponent.

File: sip-test-kit/pom.xml (hunk 1 above)
Comment (nikolag-ikor) [reply sip-test-kit/pom.xml:78]: Resolved

File: sip-test-kit/pom.xml (hunk 2 above)
Comment (vladiIkor) [reply sip-test-kit/pom.xml]: Looks to me  its available as transitive dependency of camel-spring-boot-starter and we can remove it from pom

File: sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    RemoteFileEndpoint<FTPFile> ftpEndpoint =
+        (RemoteFileEndpoint<FTPFile>)
+            SIPExchangeHelper.resolveEndpoint(inputExchange, camelContext);
+    Route route =
+        camelContext.getRoute(
+            (String) inputExchange.getProperty(Mock.ENDPOINT_ID_EXCHANGE_PROPERTY));
+    RemoteFileConsumer<FTPFile> ftpConsumer = (RemoteFileConsumer<FTPFile>) route.getConsumer();
+
+    Exchange ftpExchange = ftpConsumer.createExchange(true);
+    createFtpExchange(ftpExchange, ftpEndpoint.getConfiguration(), inputExchange);
```
Comment (vladiIkor) [inline sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java]: createFtpExchange seams miss leading
`fillUpMessage(ftpExchange.getMessage(), ftpEndpoint.getConfiguration(), inputExchange);`
or simply 
` ftpExchange.getMessage().setBody(inputExchange.getMessage().getBody());
    ftpExchange.getMessage().setHeaders(prepareFtpHeaders(endpointConfiguration, inputExchange));`
    here, in place

File: sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+/** Invoker class for triggering routes with Ftp, Ftps, Sftp consumers */
+@Component
+@RequiredArgsConstructor
+@Slf4j
+public class FtpRouteInvoker implements RouteInvoker {
+
+  private final CamelContext camelContext;
+
+  @Override
+  public Optional<Exchange> invoke(Exchange inputExchange) {
```
Comment (vladiIkor) [inline sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java:33]: Given that route and ftpEndpoint are used only to reach its child references (consumer and configuration) and that we reuse same approach across test kit, perhaps dedicated methods of SIPExchangeHelper could be created. 
Does it makes sense?

**Review by vladiIkor — DISMISSED** [review]

File: sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/headers/FtpExchangeHeaders.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
+package de.ikor.sip.foundation.testkit.workflow.whenphase.routeinvoker.headers;
+
+/** Enum with Camel FTP, FTPS, SFTP exchange headers which we can set programmatically */
+public enum FtpExchangeHeaders {
+  CAMEL_FILE_ABSOLUTE("CamelFileAbsolute"),
+  CAMEL_FILE_HOST("CamelFileHost"),
+  CAMEL_FILE_RELATIVE_PATH("CamelFileRelativePath"),
+  CAMEL_FILE_ABSOLUTE_PATH("CamelFileAbsolutePath"),
+  CAMEL_REMOTE_FILE_INPUT_STREAM("CamelRemoteFileInputStream");
```
Comment (LetoBukarica) [inline sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/headers/FtpExchangeHeaders.java]: Similar as we had on file, can you grab this from RemoteFileComponent.REMOTE_FILE_INPUT_STREAM if it's available here?

File: sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    RemoteFileEndpoint<FTPFile> ftpEndpoint =
+        (RemoteFileEndpoint<FTPFile>)
+            SIPExchangeHelper.resolveEndpoint(inputExchange, camelContext);
+    Route route =
+        camelContext.getRoute(
```
Comment (LetoBukarica) [inline sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java]: I think we have a helper function for this.

File: sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+import org.apache.camel.util.FileUtil;
+import org.apache.camel.util.ObjectHelper;
+import org.apache.commons.net.ftp.FTPFile;
+import org.springframework.stereotype.Component;
+
+/** Invoker class for triggering routes with Ftp, Ftps, Sftp consumers */
+@Component
+@RequiredArgsConstructor
+@Slf4j
+public class FtpRouteInvoker implements RouteInvoker {
```
Comment (LetoBukarica) [inline sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java:28]: Does it make sense to extend the FileInvoker and make the header functions protected there? It seems to me that it's behaving very similar. Maybe invoke function can be templated.

**Review by LetoBukarica — DISMISSED** [review]

File: sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java (hunk 4 above)
Comment (nikolag-ikor) [reply sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java]: Comment 1: You are right, dependency camel-util is transitive and will be removed.

Comment 2: Refactored with second approach.

File: sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/headers/FtpExchangeHeaders.java (hunk 6 above)
Comment (nikolag-ikor) [reply sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/headers/FtpExchangeHeaders.java]: Resolved as suggested.

File: sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java (hunk 7 above)
Comment (nikolag-ikor) [reply sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java]: Resolved with new Helper function.

File: sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java (hunk 8 above)
Comment (nikolag-ikor) [reply sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java:28]: Actually it doesn't, header calculations are not the same. Only headers "CamelFileLastModified" and "CamelMessageTimestamp" are calculated in same way, hence File and Ftp invokers have only one same method named "prepareOtherHeaders".

File: sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java (hunk 5 above)
Comment (nikolag-ikor) [reply sip-test-kit/src/main/java/de/ikor/sip/foundation/testkit/workflow/whenphase/routeinvoker/impl/FtpRouteInvoker.java:33]: Makes sence. Two new util methods which are similar to method resolveEndpoint are added to SIPExchangeHelper.
1. resolveRoute
2. resolveConsumer

**Review by LetoBukarica — APPROVED** [review]

**Review by vladiIkor — APPROVED** [review]

**Pull request merged**